### PR TITLE
Marketplace Download Packs

### PR DESCRIPTION
**Pull request by DeanArbel**
## Status
Ready

## Related Issues
fixes: https://github.com/demisto/etc/issues/43870

## Description
Add the option to download packs from `xsoar.pan.dev`.

## Screenshots
image
image

### CODE REVIEW

File: genMarketplace.js [hunk 1]
```diff
@@ -85,6 +85,91 @@ function createReadmeLink(listItem, itemType) {
...
+        }
+      },
+      version: "...",
+      "mandatoryCount": ...,
+      "optionalCount": ...
+    }
+  }
+  */
+
+  if (startKey in depsJson === false) {
```
[the remark this answers is not in the document]
Comment (reutshal) [reply genMarketplace.js:109]: Worth adding a comment about this condition

File: genMarketplace.js [hunk 2]
```diff
@@ -85,6 +85,91 @@ function createReadmeLink(listItem, itemType) {
...
+    travelDependencies(depsJson, startKey, firstLvlDepsJson);
+    depsJson[startKey]['mandatoryCount'] = Object.keys(depsJson[startKey]['mandatory']).length;
+    depsJson[startKey]['optionalCount'] = Object.keys(depsJson[startKey]['optional']).length;
+  }
+}
+
+function travelDependencies(depsJson, startKey, firstLvlDepsJson) {
+  // Travel over the dependencies json of a given Pack dependency type, while collecting all sub-dependecy mandatory packs
+  
+  if (startKey in depsJson === false) {
```
Comment (reutshal) [inline genMarketplace.js:119]:
```suggestion
  if !(startKey in depsJson) {
```
Same here regarding the comment :)

File: genMarketplace.js [hunk 3]
```diff
@@ -85,6 +85,91 @@ function createReadmeLink(listItem, itemType) {
...
+    depsJson[startKey] = {...firstLvlDepsJson[startKey]};
```
Comment (reutshal) [inline genMarketplace.js:120]: What happens if startKey is not present in firstLvlDepsJson?

File: genMarketplace.js [hunk 4]
```diff
@@ -85,6 +85,91 @@ function createReadmeLink(listItem, itemType) {
...
+  }
+
+  dependencyPacks = depsJson[startKey]['mandatory'];
+  for (var depKey in depsJson[startKey]['mandatory']) {
```
Comment (reutshal) [inline genMarketplace.js:124]:
```suggestion
  for (var depKey in dependencyPacks) {
```

File: genMarketplace.js [hunk 5]
```diff
@@ -85,6 +85,91 @@ function createReadmeLink(listItem, itemType) {
...
+    if (depsJson[depKey] === undefined) {
+      travelDependenciesJson(firstLvlDepsJson, depsJson, depKey);
+    }
+    
+    // fill mandatory sub-dependecies
+    for (var subDepKey in depsJson[depKey]['mandatory']) {
+      // skip if subDepKey is startKey or if subDepKey is already in node's collected mandatory
+      if (subDepKey !== startKey && (depsJson[startKey]['mandatory'][subDepKey] === undefined) ) {
```
Comment (reutshal) [inline genMarketplace.js:132]:
```suggestion
      if (subDepKey !== startKey && (dependencyPacks[subDepKey] === undefined) ) {
```

File: genMarketplace.js [hunk 6]
```diff
@@ -85,6 +85,91 @@ function createReadmeLink(listItem, itemType) {
...
+      }
+    }
+  }
+}
+
+function getMandatoryChainDependencies(packName, depsJson, exclusionSet) {
+  // collects all mandatory chained dependecies, while skipping over keys it collected along the way
+
+  res = {};
+  if (packName in depsJson === false) {  // sanity
```
Comment (reutshal) [inline genMarketplace.js:153]: What do you mean sanity? when will this happen?
```suggestion
  if !(packName in depsJson) {  // sanity
```

File: genMarketplace.js [hunk 7]
```diff
@@ -200,8 +319,18 @@ function genPackDetails() {
       }
     }
 
-    await parseContentItems()
+    await parseContentItems();
 
+    try {
+      if (pack.dependencies) {
+        if (pack.id in fullDepsJson === false) {
```
Comment (reutshal) [inline genMarketplace.js:326]:
```suggestion
        if !(pack.id in fullDepsJson) {
```

File: templates/details.hbs [hunk 8]
```diff
@@ -160,6 +160,69 @@ return (
...
+                                        <tr className="infoTableRow">
+                                            <th className="contentTableColName">Pack Name</th>
+                                            <th className="contentTableColDesc">Pack By</th>
+                                        </tr>
+                                    </thead>
+                                    {{#each dependencies.mandatory}}
+                                    <tbody>
+                                        <tr className="infoTableRow">
+                                            <td className="contentTableColName">
+                                                {{!-- TODO: Add better handling --}}
```
Comment (reutshal) [inline templates/details.hbs:180]: plz remove TODO as we decided that this solution is good enough

**Review by reutshal — CHANGES_REQUESTED** [review]
Nice work!
Please update when done, I also want to explore the preview site

**Comment by DeanArbel** [comment]
The PR is waiting for https://github.com/demisto/content/pull/16330
Should be handled by next week.

File: genMarketplace.js (hunk 6 above)
Comment (DeanArbel) [reply genMarketplace.js:153]: it won't happen - this is why it's sanity, in case something unexpected happens.

**Review by reutshal — APPROVED** [review]
Looks good! 🚀

**Pull request merged**